From ab46ee53623134f99d960105fb9f752a6bc992d1 Mon Sep 17 00:00:00 2001
From: Keunhong Park <keunhong@codeaurora.org>
Date: Tue, 17 Jul 2012 14:56:56 -0600
Subject: [PATCH] Add method that checks for dependency errors in changes

This check traverses the parents of a change and makes sure that they
all meet the following criteria:
  1. All parents are destined to the current branch and meet the
     current "outdated" check or
  2. All parents are already merged to the destination branch
  3. No parents are abandoned or are a draft

The method returns a set of errors described in the enum Change.Error.
If the set is empty then the change is valid.

Change-Id: If2b8e46b683afdaac8f61318eb46264c11d51f1f
---
 .../google/gerrit/common/data/ChangeDetail.java    |   10 +
 .../rpc/changedetail/ChangeDetailFactory.java      |    4 +
 .../com/google/gerrit/reviewdb/client/Change.java  |    8 +
 .../com/google/gerrit/server/DependencyUtil.java   |  192 ++++++++++++++++++++
 .../gerrit/server/query/change/ChangeData.java     |   90 +++++++++
 5 files changed, 304 insertions(+), 0 deletions(-)
 create mode 100644 gerrit-server/src/main/java/com/google/gerrit/server/DependencyUtil.java

diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/data/ChangeDetail.java b/gerrit-common/src/main/java/com/google/gerrit/common/data/ChangeDetail.java
index 74d1962..ad80c70 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/data/ChangeDetail.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/data/ChangeDetail.java
@@ -22,6 +22,7 @@ import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
+import java.util.Set;
 
 /** Detail necessary to display a change. */
 public class ChangeDetail {
@@ -45,6 +46,7 @@ public class ChangeDetail {
   protected PatchSet.Id currentPatchSetId;
   protected PatchSetDetail currentDetail;
   protected boolean canEdit;
+  protected Set<Change.DependencyError> dependencyErrors;
 
   public ChangeDetail() {
   }
@@ -230,4 +232,12 @@ public class ChangeDetail {
   public boolean canEdit() {
     return canEdit;
   }
+
+  public Set<Change.DependencyError> getDependencyErrors() {
+    return dependencyErrors;
+  }
+
+  public void setDependencyErrors(Set<Change.DependencyError> dependencyErrors) {
+    this.dependencyErrors = dependencyErrors;
+  }
 }
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
index 6660a3d..643ac5e 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
@@ -33,6 +33,7 @@ import com.google.gerrit.reviewdb.server.ReviewDb;
 import com.google.gerrit.server.AnonymousUser;
 import com.google.gerrit.server.ChangeUtil;
 import com.google.gerrit.server.CurrentUser;
+import com.google.gerrit.server.DependencyUtil;
 import com.google.gerrit.server.IdentifiedUser;
 import com.google.gerrit.server.ProjectUtil;
 import com.google.gerrit.server.account.AccountInfoCacheFactory;
@@ -212,6 +213,9 @@ public class ChangeDetailFactory extends Handler<ChangeDetail> {
       ChangeUtil.testMerge(opFactory, detail.getChange());
     }
 
+    detail.setDependencyErrors(DependencyUtil.findDependencyErrors(db,
+        detail.getChange(), changeControlFactory, null));
+
     final PatchSet.Id psId = detail.getChange().currentPatchSetId();
     final List<PatchSetApproval> allApprovals =
         db.patchSetApprovals().byChange(changeId).toList();
diff --git a/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/Change.java b/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/Change.java
index 61e6a97..e645cc5 100644
--- a/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/Change.java
+++ b/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/Change.java
@@ -98,6 +98,14 @@ import java.sql.Timestamp;
  * or when notice of the change submission occurs.
  */
 public final class Change {
+  public enum DependencyError {
+    ANCESTOR_NOT_CUR_BRANCH,
+    ANCESTOR_OUTDATED,
+    ANCESTOR_DRAFT,
+    ANCESTOR_ABANDONED,
+    ANCESTOR_NOT_VISIBLE
+  }
+
   public static class Id extends IntKey<com.google.gwtorm.client.Key<?>> {
     private static final long serialVersionUID = 1L;
 
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/DependencyUtil.java b/gerrit-server/src/main/java/com/google/gerrit/server/DependencyUtil.java
new file mode 100644
index 0000000..26e317d
--- /dev/null
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/DependencyUtil.java
@@ -0,0 +1,192 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+// * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following
+// disclaimer in the documentation and/or other materials provided
+// with the distribution.
+// * Neither the name of Code Aurora Forum, Inc. nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.server;
+
+import com.google.gerrit.reviewdb.client.Change;
+import com.google.gerrit.reviewdb.client.Change.Status;
+import com.google.gerrit.reviewdb.server.ReviewDb;
+import com.google.gerrit.server.project.ChangeControl;
+import com.google.gerrit.server.project.NoSuchChangeException;
+import com.google.gerrit.server.query.change.ChangeData;
+import com.google.gwtorm.server.OrmException;
+import com.google.inject.Provider;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class DependencyUtil {
+  private interface ChangeControlFactory {
+    ChangeControl get(Change change) throws NoSuchChangeException;
+  }
+
+  private static class FullChangeControlFactory implements ChangeControlFactory {
+    private ChangeControl.Factory factory;
+
+    public FullChangeControlFactory(ChangeControl.Factory factory) {
+      this.factory = factory;
+    }
+
+    @Override
+    public ChangeControl get(Change change) throws NoSuchChangeException {
+      return factory.controlFor(change);
+    }
+  }
+
+  private static class GenericChangeControlFactory implements
+      ChangeControlFactory {
+    private ChangeControl.GenericFactory genericFactory;
+    private CurrentUser currentUser;
+
+    public GenericChangeControlFactory(
+        ChangeControl.GenericFactory genericFactory, CurrentUser currentUser) {
+      this.genericFactory = genericFactory;
+      this.currentUser = currentUser;
+    }
+
+    @Override
+    public ChangeControl get(Change change) throws NoSuchChangeException {
+      return genericFactory.controlFor(change, currentUser);
+    }
+  }
+
+  /**
+   * Returns dependency errors in the given change or its ancestors
+   *
+   * @param db to use
+   * @param change to query
+   * @param generic control factory
+   * @param currentUser
+   * @param processed the list of already processed changes
+   * @return set of dependency errors
+   * @throws OrmException
+   * @throws NoSuchChangeException
+   */
+  public static Set<Change.DependencyError> findDependencyErrors(
+      final ReviewDb db, Change change,
+      ChangeControl.GenericFactory controlFactory, CurrentUser currentUser,
+      Map<Change.Id, Set<Change.DependencyError>> processed) throws OrmException, NoSuchChangeException {
+    return findDependencyErrors(db, change, new GenericChangeControlFactory(
+        controlFactory, currentUser), processed);
+  }
+
+  /**
+   * Returns dependency errors in the given change or its ancestors
+   *
+   * @param db to use
+   * @param change to query
+   * @param controlFactory
+   * @return set of dependency errors
+   * @throws OrmException
+   * @throws NoSuchChangeException
+   */
+  public static Set<Change.DependencyError> findDependencyErrors(
+      final ReviewDb db, Change change, ChangeControl.Factory controlFactory,
+      Map<Change.Id, Set<Change.DependencyError>> processed) throws OrmException, NoSuchChangeException {
+    return findDependencyErrors(db, change, new FullChangeControlFactory(
+        controlFactory), processed);
+  }
+
+  private static Set<Change.DependencyError> findDependencyErrors(
+      final ReviewDb db, Change change, ChangeControlFactory controlFactory,
+      Map<Change.Id, Set<Change.DependencyError>> processed) throws OrmException, NoSuchChangeException {
+    if (processed == null) {
+      processed = new HashMap<Change.Id, Set<Change.DependencyError>>();
+    }
+
+    Set<Change.DependencyError> result = new HashSet<Change.DependencyError>();
+
+    // If the change has already been processed skip it
+    if (processed.containsKey(change.getId())) {
+      return processed.get(change.getId());
+    }
+
+    ChangeData data = new ChangeData(change);
+    boolean allAncestorsVisible =
+        data.loadAncestorData(new Provider<ReviewDb>() {
+          @Override
+          public ReviewDb get() {
+            return db;
+          }
+        }, controlFactory.get(change));
+    boolean allParentsMerged = findErrors(data, result);
+
+    if (!allParentsMerged) {
+      if (!allAncestorsVisible) {
+        result.add(Change.DependencyError.ANCESTOR_NOT_VISIBLE);
+      }
+      for (ChangeData parentData : data.getDependsOn()) {
+        // If we have already found all possible errors break
+        if (result.size() == Change.DependencyError.values().length) {
+          return result;
+        }
+
+        Change parent = parentData.getChange();
+        result.addAll(findDependencyErrors(db, parent, controlFactory, processed));
+      }
+    }
+
+    processed.put(change.getId(), result);
+
+    return result;
+  }
+
+  private static boolean findErrors(ChangeData data,
+      Set<Change.DependencyError> result) {
+    Change change = data.getChange();
+    boolean allParentsMerged = true;
+
+    for (ChangeData parentData : data.getDependsOn()) {
+      Change parent = parentData.getChange();
+
+      if (parent.getStatus() == Status.MERGED) {
+        continue;
+      }
+
+      allParentsMerged = false;
+
+      if (!parent.getDest().getShortName()
+          .equals(change.getDest().getShortName())) {
+        result.add(Change.DependencyError.ANCESTOR_NOT_CUR_BRANCH);
+      }
+      if (!parent.currPatchSetId().equals(
+          data.getAncestorPatchId(parent.getId()))) {
+        result.add(Change.DependencyError.ANCESTOR_OUTDATED);
+      }
+      if (parent.getStatus() == Change.Status.DRAFT) {
+        result.add(Change.DependencyError.ANCESTOR_DRAFT);
+      }
+      if (parent.getStatus() == Change.Status.ABANDONED) {
+        result.add(Change.DependencyError.ANCESTOR_ABANDONED);
+      }
+    }
+
+    return allParentsMerged;
+  }
+}
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/query/change/ChangeData.java b/gerrit-server/src/main/java/com/google/gerrit/server/query/change/ChangeData.java
index d6762db..d346abd 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/query/change/ChangeData.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/query/change/ChangeData.java
@@ -16,17 +16,22 @@ package com.google.gerrit.server.query.change;
 
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
+import com.google.gerrit.common.data.ChangeInfo;
+import com.google.gerrit.reviewdb.client.Account;
 import com.google.gerrit.reviewdb.client.Change;
 import com.google.gerrit.reviewdb.client.ChangeMessage;
 import com.google.gerrit.reviewdb.client.Patch;
 import com.google.gerrit.reviewdb.client.PatchLineComment;
 import com.google.gerrit.reviewdb.client.PatchSet;
+import com.google.gerrit.reviewdb.client.PatchSetAncestor;
+import com.google.gerrit.reviewdb.client.RevId;
 import com.google.gerrit.reviewdb.client.PatchSet.Id;
 import com.google.gerrit.reviewdb.client.PatchSetApproval;
 import com.google.gerrit.reviewdb.client.Project;
 import com.google.gerrit.reviewdb.client.TrackingId;
 import com.google.gerrit.reviewdb.server.ReviewDb;
 import com.google.gerrit.server.CurrentUser;
+import com.google.gerrit.server.IdentifiedUser;
 import com.google.gerrit.server.git.GitRepositoryManager;
 import com.google.gerrit.server.patch.PatchList;
 import com.google.gerrit.server.patch.PatchListCache;
@@ -47,9 +52,12 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 public class ChangeData {
   public static void ensureChangeLoaded(
@@ -117,6 +125,9 @@ public class ChangeData {
   private CurrentUser visibleTo;
   private ChangeControl changeControl;
   private List<ChangeMessage> messages;
+  private ArrayList<ChangeData> dependsOn;
+  private ArrayList<ChangeData> neededBy;
+  private HashMap<Change.Id,PatchSet.Id> ancestorPatchIds;
 
   public ChangeData(final Change.Id id) {
     legacyId = id;
@@ -322,4 +333,83 @@ public class ChangeData {
     }
     return messages;
   }
+
+  public List<ChangeData> getDependsOn() {
+    return dependsOn;
+  }
+
+  public List<ChangeData> getNeededBy() {
+    return neededBy;
+  }
+
+  public PatchSet.Id getAncestorPatchId(Change.Id ck) {
+    return ancestorPatchIds.get(ck);
+  }
+
+  public boolean loadAncestorData(Provider<ReviewDb> dbProvider, ChangeControl control) throws OrmException {
+    boolean retval = true;
+
+    ensureCurrentPatchSetLoaded(dbProvider, new ArrayList<ChangeData>(Arrays.asList(this)));
+
+    ReviewDb db = dbProvider.get();
+
+    ResultSet<PatchSetAncestor> ancestors = db.patchSetAncestors().ancestorsOf(currentPatchSet.getId());
+
+    final HashSet<Change.Id> changesToGet = new HashSet<Change.Id>();
+    final HashMap<Change.Id,PatchSet.Id> ancestorPatchIds =
+        new HashMap<Change.Id,PatchSet.Id>();
+    final List<Change.Id> ancestorOrder = new ArrayList<Change.Id>();
+    for (PatchSetAncestor a : ancestors) {
+      for (PatchSet p : db.patchSets().byRevision(a.getAncestorRevision())) {
+        final Change.Id ck = p.getId().getParentKey();
+        if (changesToGet.add(ck)) {
+          ancestorPatchIds.put(ck, p.getId());
+          ancestorOrder.add(ck);
+        }
+      }
+    }
+
+    final Set<PatchSet.Id> descendants = new HashSet<PatchSet.Id>();
+    RevId cprev;
+    for (PatchSet p : patches) {
+      cprev = p.getRevision();
+      if (cprev != null) {
+        for (PatchSetAncestor a : db.patchSetAncestors().descendantsOf(cprev)) {
+          if (descendants.add(a.getPatchSet())) {
+            changesToGet.add(a.getPatchSet().getParentKey());
+          }
+        }
+      }
+    }
+    final Map<Change.Id, Change> m =
+        db.changes().toMap(db.changes().get(changesToGet));
+
+    final ArrayList<ChangeData> dependsOn = new ArrayList<ChangeData>();
+    for (final Change.Id a : ancestorOrder) {
+      final Change ac = m.get(a);
+      if (ac != null && ac.getProject().equals(change.getProject())) {
+        dependsOn.add(new ChangeData(ac));
+        if (!control.isVisible(db)) {
+          retval = false;
+        }
+      }
+    }
+
+    final ArrayList<ChangeData> neededBy = new ArrayList<ChangeData>();
+    for (final PatchSet.Id a : descendants) {
+      final Change ac = m.get(a.getParentKey());
+      if (ac != null && ac.currentPatchSetId().equals(a)) {
+        neededBy.add(new ChangeData(ac));
+        if (control.isVisible(db)) {
+          retval = false;
+        }
+      }
+    }
+
+    this.dependsOn = dependsOn;
+    this.neededBy = neededBy;
+    this.ancestorPatchIds = ancestorPatchIds;
+
+    return retval;
+  }
 }
-- 
1.7.8.3

